README.md
# MartianAtWork Laravel

[![Build Status](https://travis-ci.com/kksoftwareag/laravel-model-generator.svg?branch=master)](https://travis-ci.com/kksoftwareag/laravel-model-generator)
[![Latest Stable Version](https://poser.pugx.org/kksoftwareag/laravel-model-generator/v/stable)](https://packagist.org/packages/kksoftwareag/laravel-model-generator)
[![Total Downloads](https://poser.pugx.org/kksoftwareag/laravel-model-generator/downloads)](https://packagist.org/packages/kksoftwareag/laravel-model-generator)
[![Latest Unstable Version](https://poser.pugx.org/kksoftwareag/laravel-model-generator/v/unstable)](https://packagist.org/packages/kksoftwareag/laravel-model-generator)
[![License](https://poser.pugx.org/kksoftwareag/laravel-model-generator/license)](https://packagist.org/packages/kksoftwareag/laravel-model-generator)

MartianAtWork Laravel is a collection of Laravel Components which aim is 
to help the development process of Laravel applications by 
providing some convenient code-generation capabilities.

## How does it work?

This package expects that you are using Laravel 5.1 or above.
You will need to import the `MartianAtWork/laravel` package via composer:

```shell
composer require MartianAtWork/laravel
```

### Configuration

Add the service provider to your `config/app.php` file within the `providers` key:

```php
// ...
'providers' => [
    /*
     * Package Service Providers...
     */

    MartianAtWork\Coders\CodersServiceProvider::class,
],
// ...
```
### Configuration for local environment only

If you wish to enable generators only for your local environment, you should install it via composer using the --dev option like this:

```shell
composer require MartianAtWork/laravel --dev
```

Then you'll need to register the provider in `app/Providers/AppServiceProvider.php` file.

```php
public function register()
{
    if ($this->app->environment() == 'local') {
        $this->app->register(\MartianAtWork\Coders\CodersServiceProvider::class);
    }
}
```

## Models

![Generating models with artisan](https://cdn-images-1.medium.com/max/800/1*hOa2QxORE2zyO_-ZqJ40sA.png "Making artisan code my Eloquent models")

Add the `models.php` configuration file to your `config` directory and clear the config cache:

```shell
php artisan vendor:publish --tag=MartianAtWork-models
php artisan config:clear
```

### Usage

Assuming you have already configured your database, you are now all set to go.

- Let's scaffold some of your models from your default connection.

```shell
php artisan code:models
```

- You can scaffold a specific table like this:

```shell
php artisan code:models --table=users
```

- You can also specify the connection:

```shell
php artisan code:models --connection=mysql
```

- If you are using a MySQL database, you can specify which schema you want to scaffold:

```shell
php artisan code:models --schema=shop
```

### Customizing Model Scaffolding

To change the scaffolding behaviour you can make `config/models.php` configuration file
fit your database needs. [Check it out](https://github.com/MartianAtWork/laravel/blob/master/config/models.php) ;-)

### Tips

#### 1. Keeping model changes

You may want to generate your models as often as you change your database. In order
not to lose you own model changes, you should set `base_files` to `true` in your `config/models.php`.

When you enable this feature your models will inherit their base configurations from
base models. You should avoid adding code to your base models, since you
will lose all changes when they are generated again.

> Note: You will end up with two models for the same table and you may think it is a horrible idea 
to have two classes for the same thing. However, it is up to you
to decide whether this approach gives value to your project :-)

#### Support

For the time being, this package only supports MySQL databases. Support for other databases will be added soon.


CHANGELOG.md
# Release Notes

## 1.0.0 (2019-12-13)

## Changes
- Support Laravel 6
- Change minimum php requirement to 7.2
- Adopt new versioning scheme

## v0.0.13 (2017-02-04)

### Fixed
 - Error when using setFetchMode on Laravel Connection [#13](https://github.com/MartianAtWork/laravel/issues/11)

## v0.0.12 (2017-01-08)

### Fixed
 - Error when table with relationships does not have a primary key [#11](https://github.com/MartianAtWork/laravel/issues/11)

## v0.0.11 (2016-11-01)

### Added
 - CHANGELOG and LICENSE files

### Fixed
 - Only retrieve details for tables and not views ([#6](https://github.com/MartianAtWork/laravel/pull/6))


## v0.0.10 (2016-11-01)

### Fixed
 - Add missing backticks to database queries ([#5](https://github.com/MartianAtWork/laravel/issues/5), [#2](https://github.com/MartianAtWork/laravel/issues/2))


## v0.0.9 (2016-10-23)

### Changed
 - Treat Soft Deletes field as string


## v0.0.8 (2016-10-23)

### Fixed
 - Table name plural form resolution ([#3](https://github.com/MartianAtWork/laravel/issues/3))


## v0.0.7 (2016-10-22)

### Added
 - Travis file and badge

### Changed
 - Extended Laravel compatibility from 5.1 above.


## v0.0.6 (2016-10-20)

### Added
 - StyleCI badge

### Changed
 - Updated composer dependencies


## v0.0.5 (2016-10-16)

### Changed
 - Stop casting Soft Deletes field as date


## v0.0.4 (2016-10-16)

### Changed
 - Changed Package name in comment


## v0.0.3 (2016-10-16)

### Changed
 - Optional model hints


## v0.0.2 (2016-10-16)

### Changed
 - Updated description

## v0.0.1 (2016-10-16)

### Added
 - Created Project



src/Meta/Sqlite/Column.php
<?php
/**
 * Created by Cristian.
 * Date: 18/09/16 08:36 PM.
 */

namespace MartianAtWork\Meta\Sqlite;

use Illuminate\Support\Fluent;

class Column implements \MartianAtWork\Meta\Column
{
    /**
     * @var array
     */
    public static $mappings = [
        'string' => ['varchar', 'text', 'string', 'char', 'enum', 'tinytext', 'mediumtext', 'longtext'],
        'date' => ['datetime', 'year', 'date', 'time', 'timestamp'],
        'int' => ['bigint', 'int', 'integer', 'tinyint', 'smallint', 'mediumint'],
        'float' => ['float', 'decimal', 'numeric', 'dec', 'fixed', 'double', 'real', 'double precision'],
        'boolean' => ['longblob', 'blob', 'bit'],
    ];
    /**
     * @var \Doctrine\DBAL\Schema\Column
     */
    protected $metadata;
    /**
     * @var array
     */
    protected $metas = [
        'type', 'name', 'autoincrement', 'nullable', 'default', 'comment',
    ];

    /**
     * MysqlColumn constructor.
     *
     * @param array $metadata
     */
    public function __construct($metadata = []) {
        $this->metadata = $metadata;
    }

    /**
     * @return Fluent
     */
    public function normalize() {
        $attributes = new Fluent();
        foreach ($this->metas as $meta) {
            $this->{'parse' . ucfirst($meta)}($attributes);
        }
        return $attributes;
    }

    /**
     * @param Fluent $attributes
     */
    protected function parseType(Fluent $attributes) {
        $dataType = $this->metadata->getType()->getName();
        foreach (static::$mappings as $phpType => $database) {
            if (in_array($dataType, $database)) {
                $attributes['type'] = $phpType;
            }
        }
        if ($attributes['type'] == 'int') {
            $attributes['unsigned'] = $this->metadata->getUnsigned();
        }
    }

    /**
     * @param Fluent $attributes
     */
    protected function parseName(Fluent $attributes) {
        $attributes['name'] = $this->metadata->getName();
    }

    /**
     * @param Fluent $attributes
     */
    protected function parseAutoincrement(Fluent $attributes) {
        $attributes['autoincrement'] = $this->metadata->getAutoincrement();
    }

    /**
     * @param Fluent $attributes
     */
    protected function parseNullable(Fluent $attributes) {
        $attributes['nullable'] = $this->metadata->getNotnull();
    }

    /**
     * @param Fluent $attributes
     */
    protected function parseDefault(Fluent $attributes) {
        $attributes['default'] = $this->metadata->getDefault();
    }

    /**
     * @param Fluent $attributes
     */
    protected function parseComment(Fluent $attributes) {
        $attributes['comment'] = $this->metadata->getComment();
    }
}
